### ISSUE TITLE
[Regression] javax.xml.parsers.FactoryConfigurationError: Provider org.apache.xerces.jaxp.SAXParserFactoryImpl not found

### ISSUE THREAD

**Issue by fhossfel** [apache/camel-k#4266]
I am not sure what changed but since yesterday I am suddenly getting an error when trying to run a saxon-xslt:

    // camel-k: resource=file:common_json2xml.xslt
    
    from('timer:saxontest?period=10000')
      .setBody(constant("<data>{\"name\":\"John\", \"age\":30, \"car\":null}</data>"))
      .to('xslt-saxon:file:/etc/camel/resources/common_json2xml.xslt')
      .to("log:info?showBody=true&showHeaders=true") `

The XSLT `common_json2xml.xslt` has the following content:

    <?xml version="1.0" encoding="UTF-8"?>
    <xsl:stylesheet version="3.0" 
                    xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
                    xmlns:xs="http://www.w3.org/2001/XMLSchema"
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                    xmlns:fn="http://www.w3.org/2005/xpath-functions"
                    xsi:schemaLocation="http://www.w3.org/1999/XSL/Transform https://www.w3.org/2007/schema-for-xslt20.xsd"
                    exclude-result-prefixes="xs fn xsi">
    
        <xsl:output method="xml" encoding="UTF-8" byte-order-mark="no" omit-xml-declaration="yes" indent="yes"  />
    
        <xsl:template match="/data">
            <xsl:copy-of select="json-to-xml(.)"/>
        </xsl:template>
    
     </xsl:stylesheet>

I get the following error:
```
[1] 2023-04-21 09:00:38,513 ERROR [io.qua.run.Application] (main) Failed to start application (with profile [prod]): javax.xml.parsers.FactoryConfigurationError: Provider org.apache.xerces.jaxp.SAXParserFactoryImpl not found
[1]     at javax.xml.parsers.SAXParserFactory.newInstance(Unknown Source)
[1]     at org.xmlresolver.ResolverFeature.lambda$static$0(ResolverFeature.java:408)
[1]     at org.xmlresolver.cache.ResourceCache.reset(ResourceCache.java:208)
[1]     at org.xmlresolver.cache.ResourceCache.<init>(ResourceCache.java:145)
[1]     at org.xmlresolver.XMLResolverConfiguration.getFeature(XMLResolverConfiguration.java:1098)
[1]     at org.xmlresolver.CatalogResolver.<init>(CatalogResolver.java:51)
[1]     at org.xmlresolver.Resolver.<init>(Resolver.java:68)
[1]     at net.sf.saxon.lib.CatalogResourceResolver.<init>(CatalogResourceResolver.java:46)
[1]     at net.sf.saxon.Configuration.init(Configuration.java:367)
[1]     at net.sf.saxon.Configuration.<init>(Configuration.java:230)
[1]     at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[1]     at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[1]     at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[1]     at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
[1]     at java.base/java.lang.Class.newInstance(Class.java:584)
[1]     at net.sf.saxon.Configuration.newConfiguration(Configuration.java:246)
[1]     at net.sf.saxon.s9api.Processor.<init>(Processor.java:69)
[1]     at net.sf.saxon.jaxp.SaxonTransformerFactory.<init>(SaxonTransformerFactory.java:59)
[1]     at net.sf.saxon.TransformerFactoryImpl.<init>(TransformerFactoryImpl.java:42)
[1]     at org.apache.camel.component.xslt.saxon.XsltSaxonEndpoint.createXsltBuilder(XsltSaxonEndpoint.java:202)
[1]     at org.apache.camel.component.xslt.saxon.XsltSaxonEndpoint.doInit(XsltSaxonEndpoint.java:168)
```
This was still working yesterday morning. Not sure what has changed. I have not upgraded Camel-K. Bot CLI and operator are a 1.12.0.

**Comment by fhossfel** [comment]
For what it is worth: This is the dependency list from the build log

{"level":"info","ts":1682067527.7278018,"logger":"camel-k.builder","msg":"artifacts: [camel-k-integration-1.12.0.jar io.github.crac.org-crac-0.1.3.jar io.quarkus.quarkus-bootstrap-runner-2.16.0.Final.jar io.quarkus.quarkus-development-mode-spi-2.16.0.Final.jar io.smallrye.common.smallrye-common-io-1.13.2.jar org.glassfish.jakarta.json-1.1.6.jar org.graalvm.sdk.graal-sdk-22.3.0.jar org.jboss.logging.jboss-logging-3.5.0.Final.jar org.jboss.logmanager.jboss-logmanager-embedded-1.0.11.jar org.wildfly.common.wildfly-common-1.5.4.Final-format-001.jar com.fasterxml.jackson.core.jackson-annotations-2.14.1.jar com.fasterxml.jackson.core.jackson-core-2.14.1.jar com.fasterxml.jackson.core.jackson-databind-2.14.1.jar com.fasterxml.jackson.dataformat.jackson-dataformat-yaml-2.14.1.jar com.fasterxml.jackson.datatype.jackson-datatype-jdk8-2.14.1.jar com.fasterxml.jackson.datatype.jackson-datatype-jsr310-2.14.1.jar com.fasterxml.jackson.module.jackson-module-parameter-names-2.14.1.jar com.github.mifmif.generex-1.0.2.jar com.squareup.okhttp3.logging-interceptor-3.14.9.jar com.squareup.okhttp3.okhttp-3.14.9.jar com.squareup.okio.okio-1.17.2.jar commons-codec.commons-codec-1.15.jar dk.brics.automaton.automaton-1.11-8.jar io.fabric8.kubernetes-client-6.3.1.jar io.fabric8.kubernetes-client-api-6.3.1.jar io.fabric8.kubernetes-httpclient-okhttp-6.3.1.jar io.fabric8.kubernetes-model-admissionregistration-6.3.1.jar io.fabric8.kubernetes-model-apiextensions-6.3.1.jar io.fabric8.kubernetes-model-apps-6.3.1.jar io.fabric8.kubernetes-model-autoscaling-6.3.1.jar io.fabric8.kubernetes-model-batch-6.3.1.jar io.fabric8.kubernetes-model-certificates-6.3.1.jar io.fabric8.kubernetes-model-common-6.3.1.jar io.fabric8.kubernetes-model-coordination-6.3.1.jar io.fabric8.kubernetes-model-core-6.3.1.jar io.fabric8.kubernetes-model-discovery-6.3.1.jar io.fabric8.kubernetes-model-events-6.3.1.jar io.fabric8.kubernetes-model-extensions-6.3.1.jar io.fabric8.kubernetes-model-flowcontrol-6.3.1.jar io.fabric8.kubernetes-model-gatewayapi-6.3.1.jar io.fabric8.kubernetes-model-metrics-6.3.1.jar io.fabric8.kubernetes-model-networking-6.3.1.jar io.fabric8.kubernetes-model-node-6.3.1.jar io.fabric8.kubernetes-model-policy-6.3.1.jar io.fabric8.kubernetes-model-rbac-6.3.1.jar io.fabric8.kubernetes-model-scheduling-6.3.1.jar io.fabric8.kubernetes-model-storageclass-6.3.1.jar io.fabric8.openshift-client-6.3.1.jar io.fabric8.openshift-client-api-6.3.1.jar io.fabric8.openshift-model-6.3.1.jar io.fabric8.openshift-model-clusterautoscaling-6.3.1.jar io.fabric8.openshift-model-config-6.3.1.jar io.fabric8.openshift-model-console-6.3.1.jar io.fabric8.openshift-model-hive-6.3.1.jar io.fabric8.openshift-model-installer-6.3.1.jar io.fabric8.openshift-model-machine-6.3.1.jar io.fabric8.openshift-model-machineconfig-6.3.1.jar io.fabric8.openshift-model-miscellaneous-6.3.1.jar io.fabric8.openshift-model-monitoring-6.3.1.jar io.fabric8.openshift-model-operator-6.3.1.jar io.fabric8.openshift-model-operatorhub-6.3.1.jar io.fabric8.openshift-model-storageversionmigrator-6.3.1.jar io.fabric8.openshift-model-tuned-6.3.1.jar io.fabric8.openshift-model-whereabouts-6.3.1.jar io.fabric8.zjsonpatch-0.3.0.jar io.quarkus.arc.arc-2.16.0.Final.jar io.quarkus.quarkus-arc-2.16.0.Final.jar io.quarkus.quarkus-core-2.16.0.Final.jar io.quarkus.quarkus-fs-util-0.0.9.jar io.quarkus.quarkus-jackson-2.16.0.Final.jar io.quarkus.quarkus-jsonp-2.16.0.Final.jar io.quarkus.quarkus-kubernetes-client-2.16.0.Final.jar io.quarkus.quarkus-kubernetes-client-internal-2.16.0.Final.jar io.quarkus.quarkus-logging-json-2.16.0.Final.jar io.quarkus.quarkus-security-2.16.0.Final.jar io.quarkus.quarkus-security-runtime-spi-2.16.0.Final.jar io.quarkus.security.quarkus-security-1.1.4.Final.jar io.smallrye.common.smallrye-common-annotation-1.13.2.jar io.smallrye.common.smallrye-common-classloader-1.13.2.jar io.smallrye.common.smallrye-common-constraint-1.13.2.jar io.smallrye.common.smallrye-common-expression-1.13.2.jar io.smallrye.common.smallrye-common-function-1.13.2.jar io.smallrye.config.smallrye-config-2.13.1.jar io.smallrye.config.smallrye-config-common-2.13.1.jar io.smallrye.config.smallrye-config-core-2.13.1.jar io.smallrye.config.smallrye-config-source-yaml-2.13.1.jar io.smallrye.reactive.mutiny-1.8.0.jar jakarta.annotation.jakarta.annotation-api-1.3.5.jar jakarta.el.jakarta.el-api-3.0.3.jar jakarta.enterprise.jakarta.enterprise.cdi-api-2.0.2.jar jakarta.inject.jakarta.inject-api-1.0.jar jakarta.interceptor.jakarta.interceptor-api-1.2.5.jar jakarta.transaction.jakarta.transaction-api-1.3.3.jar net.sf.saxon.Saxon-HE-11.4.jar org.apache.camel.camel-api-3.20.1.jar org.apache.camel.camel-base-3.20.1.jar org.apache.camel.camel-base-engine-3.20.1.jar org.apache.camel.camel-bean-3.20.1.jar org.apache.camel.camel-cloud-3.20.1.jar org.apache.camel.camel-cluster-3.20.1.jar org.apache.camel.camel-componentdsl-3.20.1.jar org.apache.camel.camel-core-catalog-3.20.1.jar org.apache.camel.camel-core-engine-3.20.1.jar org.apache.camel.camel-core-languages-3.20.1.jar org.apache.camel.camel-core-model-3.20.1.jar org.apache.camel.camel-core-processor-3.20.1.jar org.apache.camel.camel-core-reifier-3.20.1.jar org.apache.camel.camel-dsl-support-3.20.1.jar org.apache.camel.camel-endpointdsl-3.20.1.jar org.apache.camel.camel-endpointdsl-support-3.20.1.jar org.apache.camel.camel-groovy-dsl-3.20.1.jar org.apache.camel.camel-groovy-dsl-common-3.20.1.jar org.apache.camel.camel-kubernetes-3.20.1.jar org.apache.camel.camel-log-3.20.1.jar org.apache.camel.camel-main-3.20.1.jar org.apache.camel.camel-management-api-3.20.1.jar org.apache.camel.camel-microprofile-config-3.20.1.jar org.apache.camel.camel-support-3.20.1.jar org.apache.camel.camel-timer-3.20.1.jar org.apache.camel.camel-tooling-model-3.20.1.jar org.apache.camel.camel-util-3.20.1.jar org.apache.camel.camel-util-json-3.20.1.jar org.apache.camel.camel-xml-io-util-3.20.1.jar org.apache.camel.camel-xml-jaxp-3.20.1.jar org.apache.camel.camel-xslt-3.20.1.jar org.apache.camel.camel-xslt-saxon-3.20.1.jar org.apache.camel.k.camel-k-core-1.17.0.jar org.apache.camel.k.camel-k-core-api-1.17.0.jar org.apache.camel.k.camel-k-core-support-1.17.0.jar org.apache.camel.k.camel-k-runtime-1.17.0.jar org.apache.camel.quarkus.camel-quarkus-bean-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-core-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-core-cloud-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-groovy-dsl-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-kubernetes-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-log-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-support-bouncycastle-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-support-commons-logging-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-support-dsl-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-timer-2.16.0.jar org.apache.camel.quarkus.camel-quarkus-xslt-saxon-2.16.0.jar org.apache.commons.commons-compress-1.22.jar org.apache.httpcomponents.client5.httpclient5-5.1.3.jar org.apache.httpcomponents.core5.httpcore5-5.1.3.jar org.apache.httpcomponents.core5.httpcore5-h2-5.1.3.jar org.bouncycastle.bcpkix-jdk15on-1.70.jar org.bouncycastle.bcprov-jdk15on-1.70.jar org.bouncycastle.bcutil-jdk15on-1.70.jar org.codehaus.groovy.groovy-3.0.8.jar org.eclipse.microprofile.config.microprofile-config-api-2.0.1.jar org.eclipse.microprofile.context-propagation.microprofile-context-propagation-api-1.2.jar org.javassist.javassist-3.26.0-GA.jar org.jboss.logging.commons-logging-jboss-logging-1.0.0.Final.jar org.jboss.logging.jboss-logging-annotations-2.2.1.Final.jar org.jboss.slf4j.slf4j-jboss-logmanager-1.2.0.Final.jar org.jboss.spec.javax.xml.bind.jboss-jaxb-api_2.3_spec-2.0.0.Final.jar org.jboss.threads.jboss-threads-3.4.3.Final.jar org.reactivestreams.reactive-streams-1.0.3.jar org.slf4j.slf4j-api-1.7.36.jar org.xmlresolver.xmlresolver-4.4.3-data.jar org.xmlresolver.xmlresolver-4.4.3.jar org.yaml.snakeyaml-1.33.jar xml-apis.xml-apis-1.4.01.jar generated-bytecode.jar quarkus-application.dat quarkus-app-dependencies.txt quarkus-run.jar]"}

**Comment by squakez** [comment]
It seems to me that the dependency that contains `org.apache.xerces.jaxp.SAXParserFactoryImpl` is missing. How are your running the integration? Is it possible that you inadvertently remove that dependency from the Integration specification. I suggest you to include it and re-run to see if it fixes.

**Comment by lfabriko** [comment]
I reproduced the original issue on Minikube (via `kamel run myintegration.groovy`)

**Comment by fhossfel** [comment]
No, I have not removed any dependency. I don't even know how do that.

I run the integration like @lfabriko did: `kamel run myintegration.groovy --dev`

I can try to reinstall the operator and see if the problem persists.

**Comment by squakez** [comment]
Please, include the dependency that contain the missing class with `kamel run -d` option.

**Comment by fhossfel** [comment]
Than I am completely clueless. But since the error will occur upon any xxslt-saxon usage it would surprise me if I was the first one to notice.

**Comment by oscerd** [comment]
The base image is always the same. It has been pushed during the release. Nobody has pushed since the release with that tag.

**Comment by fhossfel** [comment]
We have analyzed the issue further:

1.) The described workaround works for the xslt-saxon component but it does not work for the xtokenize splitter. I will complain about a missing XPathFactory. I have not found out which dependency to add to fix that.

2.) This is related to the upgrade of Saxon from 9.9 (?) to version 11.4 during the 1.12. We have currently downgraded to Camel-K version 1.11.1 which ships with Saxon 9.9 and xslt-saxon and tokenize both work in that version.

However, we need XSLT 3.0 features (xsl:evaluate) so we are currently adding a dependency on net.sf.saxon:Saxon-HE:10.9. This works for the moment but is a bit hackish. If I do find the time I will upgrade Camel-K to 1.12 again and see if I can downgrade Saxon to 10.9.

**Comment by squakez** [comment]
Hi, thanks for the detailed information. The way the dependecies are managed mainly depends on the way they are transitively brought in by the camel component you're using. Ideally, when using a component, you need to verify in the Camel documentation which is the way to configure it and, if it does require some additional dependency. If you notice some problem, I suggest you verify how it works in a plain Camel application and report that issue to the Camel framework directly. Once this is solved in Camel, it will be eventually solved in Camel K.

**Comment by fhossfel** [comment]
I am 99% positive this not related to the dependencies in the camel component but to the old JAXP way to instantiate the XML processor/XSLT factory. JAXP relies on some really weird reflection stuff to instantiate the XML processing components. This seems to be a source of problems for Quarkus (see quarks/quarkus #1769)[https://github.com/quarkusio/quarkus/issues/1762].

I am not sure why Saxon version 10 works but version 11 doesn't but I assume that some JAXP related stuff was removed from Saxon 11. Or Quarkus since tokenize seems to be affected, too.

**Comment by fhossfel** [comment]
Has anybody looked into this? XML-Processing does not work in Camel-K until this is fixed.

If there is anything I can do, plz let me know.

**Comment by squakez** [comment]
As we suspect the problem is on the runtime (Camel Quarkus), I think the best thing to do is try to replicate the issue with a local Camel Quarkus application using the same component. If you reproduce the issue, you have material to share in the Camel Quarkus project for further analysis/fix.

**Comment by fhossfel** [comment]
Just as a note for anyone who stumbles on this: This has definitely been fixed in 2.10 - which ships with Saxon HE 12.3.

This may have been fixed by https://github.com/apache/camel-quarkus/issues/4065
